mungegithub/mungers/triage-filer.go

```diff
@@ -0,0 +1,562 @@
+	row, err := jdata.rowMap.rowForBuild(buildnum)
+	if err != nil {
+		return 0, err
+	}
+	if row < 0 || row >= len(jdata.starts) {
+		return 0, fmt.Errorf("Invalid row index for jobpath '%s'. Index is %d, should be in [0,%d).", jdata.jobPath, row, len(jdata.starts))
+	}
+	return jdata.starts[row], nil
+}
+
+//topTestsFailing returns the top 'count' test names sorted by number of failing jobs.
+func (clust *Cluster) topTestsFailed(count int) []string {
+	tests := make([]interface{}, len(clust.testsFailed))
+	j := 0
+	for testName, _ := range clust.testsFailed {
+		tests[j] = testName
+		j++
+	}
+	sorter := &slicesorter{slice: tests,
+		less: func(a, b int) bool {
+			return len(clust.testsFailed[tests[a].(string)]) > len(clust.testsFailed[tests[b].(string)])
+		},
+	}
+
+	sort.Sort(sorter)
+
+	if len(tests) < count {
+		count = len(tests)
+	}
+	result := make([]string, count)
+	for i := 0; i < count; i++ {
+		result[i] = tests[i].(string)
+	}
+	return result
+}
+
+//topJobsFailed returns the top 'count' job names sorted by number of failing builds.
+func (clust *Cluster) topJobsFailed(count int) []string {
+	jobs := make([]interface{}, len(clust.jobsFailed))
+	j := 0
+	for jobName, _ := range clust.jobsFailed {
+		jobs[j] = jobName
+		j++
+	}
+	sorter := &slicesorter{slice: jobs,
+		less: func(a, b int) bool {
+			return len(clust.jobsFailed[jobs[a].(string)].builds) > len(clust.jobsFailed[jobs[b].(string)].builds)
+		},
+	}
+
+	sort.Sort(sorter)
+
+	if len(jobs) < count {
+		count = len(jobs)
+	}
+	result := make([]string, count)
+	for i := 0; i < count; i++ {
+		result[i] = jobs[i].(string)
+	}
+	return result
+}
+
+//The following funcs implement the Issue interface
+
+//Title is the string to use as the github issue title.
+func (clust *Cluster) Title() string {
+	return fmt.Sprintf("Cluster [%s...] failed %d builds, %d jobs, and %d tests over %d days",
+		clust.keyHash[0:6],
+		clust.totalBuilds,
+		clust.totalJobs,
+		clust.totalTests,
+		clust.filer.windowDays,
+	)
+}
+
+//Body returns the body text of the github issue and *must* contain the output of ID().
+//closedIssues is a (potentially empty) slice containing all closed issues authored by this bot
+//that contain ID() in their body.
+//If Body returns an empty string no issue is created.
+func (clust *Cluster) Body(closedIssues []*github.MungeObject) string {
+	//first check that the most recently closed issue (if any exist) was closed
+	//before the start of the sliding window
+	cutoffTime := time.Now().AddDate(0, 0, -1*clust.filer.windowDays)
+	for _, closed := range closedIssues {
+		if closed.Issue.ClosedAt.After(cutoffTime) {
+			return ""
+		}
+	}
+
+	//now we know we need to make a new issue
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "#### Cluster [%s] More data at: %s#%s\n", clust.ID(), clust.filer.triageURL, clust.ID())
+	fmt.Fprintf(&buf, "Failure stats over time range '%s' to '%s'\n", cutoffTime.Format(time.RFC822), time.Now().Format(time.RFC822))
```

Comment:
again, the time range should be specified based on what's *in the data file*, not the current time.